unittests/Sema/KeypathFunctionConversionTests.cpp

```diff
@@ -0,0 +1,127 @@
+/// (e.g., the overload *is* generic).
+TEST_F(SemaTest, TestKeypathFunctionConversionPrefersNarrowConversion) {
+  auto *stringTypeDecl = getStdlibNominalTypeDecl("String");
+  auto *boolTypeDecl = getStdlibNominalTypeDecl("Bool");
+  auto boolType = boolTypeDecl->getDeclaredType();
+  auto boolOptType = OptionalType::get(boolType);
+  auto stringType = stringTypeDecl->getDeclaredType();
+
+  auto voidType = TupleType::get({}, Context);
+
+  auto *genericParam1 = new (Context) GenericTypeParamDecl(
+      DC, Context.getIdentifier("T"), SourceLoc(), SourceLoc(), 0, 0, false);
+  auto genericType1 =
+      genericParam1->getDeclaredInterfaceType()->getAs<GenericTypeParamType>();
+
+  auto *genericParam2 = new (Context) GenericTypeParamDecl(
+      DC, Context.getIdentifier("U"), SourceLoc(), SourceLoc(), 0, 1, false);
+  auto genericType2 =
+      genericParam2->getDeclaredInterfaceType()->getAs<GenericTypeParamType>();
+
+  auto declName = DeclName(Context, Context.getIdentifier("f"), {Identifier()});
+
+  // func f<T, U>(_: (T) -> U))
+  auto innerGenericFnParam = AnyFunctionType::Param(genericType1);
+  auto genericFnParamTy = FunctionType::get({innerGenericFnParam}, genericType2)
+                              ->withExtInfo(AnyFunctionType::ExtInfo());
+  auto *genericFnParamDecl = new (Context) ParamDecl(
+      SourceLoc(), SourceLoc(), Identifier(), SourceLoc(), Identifier(), DC);
+  genericFnParamDecl->setSpecifier(ParamSpecifier::Default);
+  auto *genericFnParamList = ParameterList::create(
+      Context, SourceLoc(), {genericFnParamDecl}, SourceLoc());
+  llvm::SmallVector<GenericTypeParamDecl *, 2> genericParams = {genericParam1,
+                                                                genericParam2};
+  auto *genericParamList =
+      GenericParamList::create(Context, SourceLoc(), {}, SourceLoc());
+  auto genericFnDecl = FuncDecl::create(
+      Context, SourceLoc(), StaticSpellingKind::None, SourceLoc(), declName,
+      SourceLoc(), /*async=*/false, SourceLoc(), /*throws=*/false, SourceLoc(),
+      nullptr, genericParamList, genericFnParamList, nullptr, DC);
+  auto genericFnParam = AnyFunctionType::Param(genericFnParamTy);
+  llvm::SmallVector<GenericTypeParamType *, 2> genericTypeParams = {
+      genericType1, genericType2};
+  auto genericSig = GenericSignature::get(genericTypeParams, {});
+  auto genericFnTy =
+      GenericFunctionType::get(genericSig, {genericFnParam}, voidType)
+          ->withExtInfo(AnyFunctionType::ExtInfo());
+  genericFnDecl->setInterfaceType(genericFnTy);
+
+  // func f(_: (String) -> Bool?)
+  auto innerConcreteFnParam = AnyFunctionType::Param(stringType);
+  auto concreteFnParamTy =
+      FunctionType::get({innerConcreteFnParam}, boolOptType)
+          ->withExtInfo(AnyFunctionType::ExtInfo());
+  auto *concreteFnParamDecl = new (Context) ParamDecl(
+      SourceLoc(), SourceLoc(), Identifier(), SourceLoc(), Identifier(), DC);
```

Comment:
There are means to create ParamDecl implicitly which don't require all these empty source locations and identifiers.